### DISCUSSION TITLE
serverless.ts is not handling ${self:...} syntax

### DISCUSSION THREAD

**Discussion by jeremycare** [serverless/serverless#10856] [Q&A] [answered]
Do there is a specific reason why the `${opt:..}` syntax works fine when the configuration is in typescript (serverless.ts) and `${self:...}` is not interpreted correctly?

`TableName: 'RATES_TABLE_${opt:stage}'` works as expected.
`TableName: 'RATES_TABLE_${self:provider.stage}',` error out.

Any idea?

**Comment by pgrzesik** [comment, marked as answer]
Hello @jeremycare - could you provide an error message you're getting? Is the `provider.stage` actually defined in your configuration file? If not, then it will rightfully error out - it's the same behavior for `.ts` and `.yml` configurations.

**Comment by jeremycare** [comment]
You're totally right, `provider.stage` was not defined.

I was trying to import the value in a custom variable,  as follow:
```
    custom: {
      RATES_TABLE: "RATES_TABLE_${opt:stage, self:provider.stage, 'dev'}",
    }
```
and was getting:
`  - Cannot resolve variable at "custom.ratesTable": Value not found at "self" source`

when I was using it in a resource:
`          TableName: '${self:custom.ratesTable}',`

So I thought it was not getting the self values correctly.

Sorry about that

**Reply by pgrzesik** [reply]
No problem, happy you were able to resolve your issue 👍

**Reply by Markkos89** [reply]
> You're totally right, `provider.stage` was not defined.
> 
> I was trying to import the value in a custom variable, as follow:
> 
> ```
>     custom: {
>       RATES_TABLE: "RATES_TABLE_${opt:stage, self:provider.stage, 'dev'}",
>     }
> ```
> 
> and was getting: ` - Cannot resolve variable at "custom.ratesTable": Value not found at "self" source`
> 
> when I was using it in a resource: ` TableName: '${self:custom.ratesTable}',`
> 
> So I thought it was not getting the self values correctly.
> 
> Sorry about that

one noob question: It does not matter the camel case or uppercase in this matter callgin self.custom ?

he defined it in uppercase and with and underscore:

```yml
custom: {
      **RATES_TABLE**: "RATES_TABLE_${opt:stage, self:provider.stage, 'dev'}",
}
```


and then calling it in camelCase.:

```yml 
TableName: '${self:custom.**ratesTable**}',
```

The answer will help me improve my configuration, thanks !!!

**Reply by mnapoli** [reply]
@Markkos89 I would expect the variables to be case sensitive, so I wouldn't expect your example to work.

**Reply by Markkos89** [reply]
you mean the example of [jeremycare](https://github.com/jeremycare) I was quoting, right?

**Reply by mnapoli** [reply]
Ah right, yes
